Select: Portal select menu to document.body

ashharrison90 opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/main/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the main branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

- portals the `Select` menu to `document.body`
- `auto` positions the `Select` menu for better placement
- closes the `Select` menu when scrolling anywhere outside of the `Select` menu
- adds a new `portal` z-index value set to higher than the modal
- rewrites `Portal` component as a functional component so it can utilise `useTheme2`
- rewrites `SelectBase` tests as RTL tests
- updates other unit tests
  - `react-select-event` has an extra option to specify where the menu is contained. e.g.
    ```javascript
    await selectEvent.select(selectEl, 'Option 2');
    ```
    becomes
    ```javascript
    await selectEvent.select(selectEl, 'Option 2', { container: document.body });
    ```
    or use the new util method exposed from `@grafana/ui`:
    ```javascript
    import { selectOptionInTest } from '@grafana/ui';
    await selectOptionInTest(selectEl, 'Option 2');
    ```
- updates e2e tests
  - as the menu is no longer within the `Select` container, need to look for options on the body. e.g.
    ```javascript
    e2e.components.TimeZonePicker.container()
      .should('be.visible')
      .within(() => {
        e2e.components.Select.input().should('be.visible').click();
        e2e.components.Select.option().should('be.visible').contains(toTimeZone).click();
      });
    ``` 
    should now be written as 
    ```javascript
    e2e.components.TimeZonePicker.container()
      .should('be.visible')
      .within(() => {
        e2e.components.Select.input().should('be.visible').click();
      });
    e2e.components.Select.option().should('be.visible').contains(toTimeZone).click();
    ```
- have also updated the e2e util method `selectOption` here: https://github.com/grafana/grafana/blob/main/packages/grafana-e2e/src/flows/selectOption.ts to work with the new structure.
- since the menu is portalled to body, wrapping the `Select` in something like `ClickOutsideWrapper` will no longer work. the same functionality can be achieved using `Select`'s `onCloseMenu`/`isOpen` props.

some screenshots:
![image](https://user-images.githubusercontent.com/20999846/125436201-86808ed6-b0f8-4a48-baee-7fe8ce6e21ee.png)
![image](https://user-images.githubusercontent.com/20999846/125436228-915e45ad-ba4f-4947-a679-f7d37d19359d.png)
![image](https://user-images.githubusercontent.com/20999846/125436281-290dca2f-aade-4c98-a0a8-1cbf561f027e.png)


**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

**Special notes for your reviewer**:

ashharrison90 committed:
ValueMappings: Force overflowing modal content to scroll

ashharrison90 committed:
ValueMappings: Update unit tests

torkelo commented:
Awesome job! this is tricky stuff, and we have had max-width / overflow hidden/auto before in the Modal but it issues (esp with selects) made us change that back. There are some other Modals with selects in them (like dashboard SaveAs modal).

Clarity-89 reviewed (approved):
Looks good! We don't have any use cases of Selects in Modal in Enterprise as far as I can see. Regarding the value mappings and modal in general, I think a nice improvement would be if the button actions wouldn't be a pat of scrolling content and always stayed visible during overflow, but this is probably out of scope of this PR.

ashharrison90 commented:
@Clarity-89 +1 on that. i think every major design library out there (bootstrap, material, carbon, etc.) separates the actions from the content for this reason.

torkelo commented:
An idea would to add a footer prop to Modal that automatically wraps the footer content in a ModalButtonRow, and the footer would live outside the content div

tskarhed commented on packages/grafana-ui/src/components/Select/types.ts:
```diff
@@ -41,6 +41,7 @@ export interface SelectCommonProps<T> {
   minMenuHeight?: number;
   maxVisibleValues?: number;
   menuPlacement?: 'auto' | 'bottom' | 'top';
+  menuPortalTarget?: HTMLElement;
```
Maybe it is okay in this case

ashharrison90 replied:
hmm, that's a good point. i'm not sure how we can avoid it though.
we could attach all selects to `document.body` by default? that would have a pretty wide impact though, and is presumably slightly worse for performance.
ideally we'd want something where the select detects it's in a modal and attaches itself to document.body automatically 🤔

tskarhed replied:
I played around a bit with the portaling etc. last year, and it never ended up good. When this is merged we might have to keep an eye out for Select bugs.

Let me just check a few places before you merge this PR.

ashharrison90 commented:
> An idea would to add a footer prop to Modal that automatically wraps the footer content in a ModalButtonRow, and the footer would live outside the content div

yep, i don't make raising that/taking it on as an enhancement following this PR 👍

tskarhed reviewed (approved):
Okay, let's go! Be on the lookout for weird Select bugs though!

ashharrison90 commented:
@tskarhed @torkelo what do you think about using context to determine whether the `Select` component is in a modal or not? 🤔 

we could export a context object from `SelectBase` - something like `IsInModalContext` - default it to false and wrap the modal like:
```jsx
    <Portal>
      <div className={cx(styles.modal, className)}>
        <IsInModalContext.Provider value={true}>
          <div className={headerClass}>
            {typeof title === 'string' && <DefaultModalHeader {...props} title={title} />}
            {typeof title !== 'string' && title}
            <div className={styles.modalHeaderClose}>
              <IconButton surface="header" name="times" size="xl" onClick={onDismiss} />
            </div>
          </div>
          <div className={cx(styles.modalContent, contentClassName)}>{children}</div>
        </IsInModalContext.Provider>
      </div>
      <div
        className={styles.modalBackdrop}
        onClick={onClickBackdrop || (closeOnBackdropClick ? onDismiss : undefined)}
      />
    </Portal>
```

that way we wouldn't have to expose any new props on `SelectBase` or `ValuePicker`.

ashharrison90 committed:
Select: Portal Select to document.body, close menu on scroll

ashharrison90 committed:
Select: Fix tests + apply updates from https://github.com/grafana/grafana/pull/32833

ashharrison90 committed:
ValueMappingsEditorModal: Revert to using selectEvent in the tests

ashharrison90 committed:
Select: Fix remaining unit tests

ashharrison90 commented:
ok, i've updated this to portal `Select` to `document.body` in all cases and updated the unit tests accordingly. the select menu now also closes when you scroll outside of the select menu.

the only thing left to discuss is what we do about z-index. in this PR the select menu overlays everything, including modals. i can imagine there's cases where that's not desirable.

but when a select is within a modal, it **has** to overlay the modal, else it's broken. so it seems like we'll need to make it configurable in some way. either by exposing a prop, or setting it via context, or something else.

torkelo commented on public/app/plugins/datasource/influxdb/components/RawInfluxQLEditor.test.tsx:
```diff
@@ -57,7 +57,7 @@ describe('RawInfluxQLEditor', () => {
     expect(formatSelect).toBeInTheDocument();
 
-    await select(formatSelect, 'Time series');
+    await select(formatSelect, 'Time series', { container: document.body });
```
this seems like a big pain to always have to specify in tests

mckn replied:
Maybe add a `select` function to grafana-ui that basically just does this by default?

torkelo commented on packages/grafana-ui/src/themes/default.ts:
```diff
@@ -123,15 +135,7 @@ const theme: GrafanaThemeCommons = {
   },
   panelPadding: 8,
   panelHeaderHeight: 28,
-  zIndex: {
-    navbarFixed: 1000,
-    sidemenu: 1020,
-    dropdown: 1030,
-    typeahead: 1030,
-    tooltip: 1040,
-    modalBackdrop: 1050,
-    modal: 1060,
-  },
+  zIndex: appZIndexes,
```
this is the old theme, new theme we introduced in v8 is here https://github.com/grafana/grafana/blob/main/packages/grafana-data/src/themes/zIndex.ts  ,  think we missed making this use properties from the new theme

torkelo commented on packages/grafana-ui/src/components/Portal/Portal.tsx:
```diff
@@ -1,5 +1,6 @@
 ﻿import React, { PureComponent } from 'react';
 import ReactDOM from 'react-dom';
+import { appZIndexes } from '../../themes/default';
```
Feels cleaner to rewrite this a a functional component and use useTheme2 to access the theme object (that has zIndex)

torkelo commented:
Awesome job on this one ash! think it's working well, the hide on scroll is also working. 

A bit afraid of this change, but it does simplify things to always portal the select. No longer any restrictions on usage or special things we need to remember when placing select inside CustomScrollbars or Modal). The impact to testing of Select is the thing that makes me a bit worried (I know Selects are used heavily in e2e tests in plugins as well). Can we introduce a util function we can reuse to share the logic of selecting a menu item? Won't solve the e2e / breaking issue but looks impossible to avoid that

ashharrison90 committed:
Portal: Rewrite Portal as a functional component so we can use useTheme2

ashharrison90 committed:
Modal: Remove modal styles from this PR

ashharrison90 committed:
Update E2E tests

ashharrison90 commented on public/app/features/alerting/unified/AmRoutes.test.tsx:
```diff
@@ -343,13 +344,13 @@ describe('AmRoutes', () => {
   await userEvent.type(inputs[0], value);
   userEvent.click(inputs[1]);
-  userEvent.click(byText(timeUnit).get(selectElement));
+  await selectEvent.select(selectElement, timeUnit, { container: document.body });
```
i could be wrong, but i believe this is a menu element it's clicking

ashharrison90 commented:
this has ended up being a much more significant change than originally intended. 

i've updated the description of the PR and removed the modal changes so this will focus solely on the `Select` portalling. i'll deliver the modal changes as part of a separate (much smaller) PR.

i've also moved the milestone to 8.1 and marked it as a breaking change. strictly speaking there's no breaking api changes, but it's a change in behaviour and definitely a breaking change for anyone relying on the old behaviour in unit/e2e tests. 

@torkelo think i've addressed your comments apart from the util function. i'll start looking into some sort of util function for the e2e side

ashharrison90 committed:
More unit test fixes

ashharrison90 committed:
Select: Fix remaining E2E tests

ashharrison90 commented:
i believe there's already an e2e method here: https://github.com/grafana/grafana/blob/ash/36317/packages/grafana-e2e/src/flows/selectOption.ts#L11

i've updated it to search for options outside of the container now 👍

mckn reviewed (approved):
Great work on this @ashharrison90! Looks really good and seems to work in the scenarios that I have tested it.

mckn commented:
Love that you included instructions on how to change the tests in this PR! Will help people that might run into this issue! 🙌🏻

torkelo reviewed (approved):
Awesome job on this ash!

only thing that can be improved would be a test helper function to select options so you not have to always pass in the container argument, should make it easier for to change this in the future without updating all tests (I hope we do no need to!!)

ashharrison90 committed:
Select: Create util method to select an option in tests

ashharrison90 commented:
unit test util method created. once everything goes green i'll merge 🤞

dprokop commented on packages/grafana-ui/src/components/Select/SelectBase.tsx:
```diff
@@ -329,19 +334,16 @@ export function SelectBase<T>({
-            position,
-            width,
-            zIndex: theme.zIndex.dropdown,
+          menuPortal: (base: any) => ({
+            ...base,
+            zIndex: theme.zIndex.portal,
```
was that it? :)

ashharrison90 replied:
oooo maybe, forgot about that change 😂
